c-dependencies/js-compute-runtime/js-compute-builtins.cpp

```diff
@@ -3625,6 +3630,388 @@ bool response_started(JSObject *self) {
+  return mozilla::Ok();
+}
+
+inline JS::Result<mozilla::Ok> base64Decode3to2(std::string_view input, std::string &output) {
+  uint8_t w, x, y;
+  // 8.1 Find the code point pointed to by position in the second column of Table 1: The Base 64
+  // Alphabet of RFC 4648. Let n be the number given in the first cell of the same row. [RFC4648]
+  if (!base64CharacterToValue(input[0], &w) || !base64CharacterToValue(input[1], &x) ||
+      !base64CharacterToValue(input[2], &y)) {
+    return JS::Result<mozilla::Ok>(JS::Error());
+  }
+  // 9. If buffer is not empty, it contains either 12 or 18 bits. If it contains 12 bits, then
+  // discard the last four and interpret the remaining eight as an 8-bit big-endian number. If it
+  // contains 18 bits, then discard the last two and interpret the remaining 16 as two 8-bit
+  // big-endian numbers. Append the one or two bytes with values equal to those one or two numbers
+  // to output, in the same order.
+  output += (uint8_t(w << 2 | x >> 4));
+  output += (uint8_t(x << 4 | y >> 2));
+  return mozilla::Ok();
+}
+
+inline JS::Result<mozilla::Ok> base64Decode2to1(std::string_view input, std::string &output) {
+  uint8_t w, x;
+  // 8.1 Find the code point pointed to by position in the second column of Table 1: The Base 64
+  // Alphabet of RFC 4648. Let n be the number given in the first cell of the same row. [RFC4648]
+  if (!base64CharacterToValue(input[0], &w) || !base64CharacterToValue(input[1], &x)) {
+    return JS::Result<mozilla::Ok>(JS::Error());
+  }
+  // 9. If buffer is not empty, it contains either 12 or 18 bits. If it contains 12 bits, then
+  // discard the last four and interpret the remaining eight as an 8-bit big-endian number. If it
+  // contains 18 bits, then discard the last two and interpret the remaining 16 as two 8-bit
+  // big-endian numbers. Append the one or two bytes with values equal to those one or two numbers
+  // to output, in the same order.
+  output += (uint8_t(w << 2 | x >> 4));
+  return mozilla::Ok();
+}
+
+bool isAsciiWhitespace(char c) {
+  switch (c) {
+  case '\t':
+  case '\n':
+  case '\f':
+  case '\r':
+  case ' ':
+    return true;
+  default:
+    return false;
+  }
+}
+
+// https://infra.spec.whatwg.org/#forgiving-base64-decode
+JS::Result<std::string> forgivingBase64Decode(std::string_view data) {
+  // 1. Remove all ASCII whitespace from data.
+  // ASCII whitespace is U+0009 TAB, U+000A LF, U+000C FF, U+000D CR, or U+0020 SPACE.
+  auto hasWhitespace = std::find_if(data.begin(), data.end(), &isAsciiWhitespace);
+  std::string dataWithoutAsciiWhitespace;
+
+  if (hasWhitespace) {
+    dataWithoutAsciiWhitespace = data;
+    dataWithoutAsciiWhitespace.erase(std::remove_if(dataWithoutAsciiWhitespace.begin(),
+                                                    dataWithoutAsciiWhitespace.end(),
+                                                    &isAsciiWhitespace),
+                                     dataWithoutAsciiWhitespace.end());
```

Comment:
Since you already know how far in the whitespace was inside the string, you can reuse the offset into the copy. Also, this will remove from the first occurrence of whitespace to the end of the string, is that correct?
```suggestion
    dataWithoutAsciiWhitespace.erase(dataWithoutAsciiWhitespace.begin() + std::distance(data.begin(), hasWhitespace),
                                     dataWithoutAsciiWhitespace.end());
```

Reply:
This should be only removing elements which return true from the `&isAsciiWhitespace` function that we pass in to `remove_if`

`remove_if` does not actually remove elements but instead moves all the elements that return false for the `&isAsciiWhitespace` function to the front of the range, leaving all the elements which return true for `&isAsciiWhitespace` at the end of the range, this is done in a single pass which is great. remote_if will then return an iterator which points to the first element that returned true for `&isAsciiWhitespace`. We then pass that returned iterator as the first parameter to `erase`, and `erase` removes all the ascii-whitespace elements in one call.

When we move to c++20 we can change from using this ['erase-remove idiom'](https://en.wikipedia.org/wiki/Erase%E2%80%93remove_idiom) to using std::remove_if instead 😄

Reply:
Sorry, I totally missed the erase/remove_if combination there, and thought it was just erasing to the end of the string!